File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int rows, cols;
    
    public int solution(String[] maps) {
        rows = maps.length;
        cols = maps[0].length();
        
        int startX = 0, startY = 0;
        int leverX = 0, leverY = 0;
        
        // 시작점과 레버 위치 찾기
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (maps[i].charAt(j) == 'S') {
                    startX = i;
                    startY = j;
                } else if (maps[i].charAt(j) == 'L') {
                    leverX = i;
                    leverY = j;
                }
            }
        }
        
        // S에서 L까지 최단 거리 찾기
        int startToLever = bfs(maps, startX, startY, 'L');
        if (startToLever == -1) {
            return -1;
        }
        
        // L에서 E까지 최단 거리 찾기
        int leverToExit = bfs(maps, leverX, leverY, 'E');
        if (leverToExit == -1) {
            return -1;
        }
        
        return startToLever + leverToExit;
    }
    
    private int bfs(String[] maps, int startX, int startY, char target) {
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[rows][cols];
        queue.add(new int[]{startX, startY, 0});
        visited[startX][startY] = true;
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            int distance = current[2];
            
            if (maps[x].charAt(y) == target) {
                return distance;
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                // 배열 범위 내에 있고 벽이 아니며 방문하지 않은 경우만 이동
                if (nx >= 0 && ny >= 0 && nx < rows && ny < cols &&
                    !visited[nx][ny] && maps[nx].charAt(ny) != 'X') {
                    visited[nx][ny] = true;
                    queue.add(new int[]{nx, ny, distance + 1});
                }
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    answer = 0
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    rows = len(maps)
    cols = len(maps[0])
    
    def bfs(i, j, target):
        queue = deque()
        queue.append((i,j,0))
        visited = [[False]*cols for _ in range(rows)]
        visited[i][j] = True
        
        while queue:
            r,c,dist = queue.popleft()
            
            if maps[r][c] == target:
                return dist
            
            for k in range(4):
                nx = dx[k] + r
                ny = dy[k] + c
                
                # 유효성 검사
                if(nx >= 0 and ny >=0 and nx < rows and ny < cols and maps[nx][ny] != 'X' and not visited[nx][ny]):
                    visited[nx][ny] = True
                    queue.append((nx, ny, dist+1))
        return -1
    
    for i in range(len(maps)):
        for j in range(len(maps[i])):
            # 출발점에서 레버까지
            if maps[i][j] == 'S':
                start_r = i
                start_c = j
                
            if maps[i][j] == 'L':
                end_r = i
                end_c = j
                
    dist_to_lever = bfs(start_r,start_c,'L')          
    dist_to_end = bfs(end_r, end_c,'E')
    
    # 둘 중 하나라도 -1이면 탈출 불가
    if dist_to_lever == -1 or dist_to_end == -1:
        answer = -1
    else:
        answer = dist_to_lever + dist_to_end
                    
    return answer